sync/statesync/code_syncer.go

```diff
@@ -145,10 +123,26 @@ func (c *codeSyncer) work(ctx context.Context) error {
 				return nil
 			}
 
+			// Skip work already present locally.
+			if rawdb.HasCode(c.db, codeHash) {
+				// Best-effort cleanup of stale marker.
+				batch := c.db.NewBatch()
+				customrawdb.DeleteCodeToFetch(batch, codeHash)
+
+				if err := batch.Write(); err != nil {
+					return fmt.Errorf("failed to write batch for stale code marker: %w", err)
+				}
+				continue
+			}
+
+			// Deduplicate in-flight code hashes across workers.
+			if _, loaded := c.inFlight.LoadOrStore(codeHash, struct{}{}); loaded {
+				continue
+			}
```

Comment:
This needs to be before the `rawdb.HasCode()` check otherwise it could race. Effectively, the `inFlight` map is an atomic way of taking responsibility for a single hash by first checking if it's already in the database and then fetching+writing if it's not. By placing the `LoadOrStore()` here, multiple workers can all hammer `HasCode()` with the same hash.